hydromt/data_sources/data_source.py

```diff
@@ -0,0 +1,94 @@
+"""Abstract DataSource class."""
+from os.path import abspath, join
+from pathlib import Path
+from typing import Any, ClassVar, Dict, Optional, Union
+
+from pydantic import BaseModel, Field, ValidationInfo, field_validator
+
+from hydromt._typing import DataType
+from hydromt.data_adapter.caching import _uri_validator
+from hydromt.metadata_resolvers import MetaDataResolver
+from hydromt.metadata_resolvers.resolver_plugin import RESOLVERS
+
+
+class DataSource(BaseModel):
+    """
+    A DataSource is a parsed section of a DataCatalog.
+
+    The DataSource, specific for a data type within HydroMT, is responsible for
+    validating the input from the DataCatalog, to
+    ensure the workflow fails as early as possible. A DataSource has information on
+    the driver that the data should be read with, and is responsible for initializing
+    this driver.
+    """
+
+    @classmethod
+    def submodel_validate(
+        cls,
+        obj: Any,
+        *,
+        strict: Optional[bool] = None,
+        from_attributes: Optional[bool] = None,
+        context: Optional[Dict[str, Any]] = None,
+    ) -> "DataSource":
+        """Validate the DataSource."""
+        if data_type := obj.get("data_type"):
+            try:
+                target_cls: DataSource = next(
+                    filter(lambda sc: sc.data_type == data_type, cls.__subclasses__())
+                )  # subclasses should be loaded from __init__.py
+                return target_cls.model_validate(
+                    obj,
+                    strict=strict,
+                    from_attributes=from_attributes,
+                    context=context,
+                )
+            except StopIteration:
+                raise ValueError(f"Unknown 'data_type': '{data_type}'")
+
+        raise ValueError("DataSource needs 'data_type'")
```

Comment:
I'm having a hard time following this code, can you elaborate a bit on what it's doing?

Reply:
I wanted to be able to construct DataSources in a generic way from it's base class, as that is what should happen when we read from the DataCatalog. This should be based on the `data_type` in the `DataCatalog` entry. In this case walk through the following steps:
1. Check whether the entry has a `data_type` field, otherwise raise an error.
2. Then try to find a subclass with a matching `data_type` field, which is a required class attribute for `DataSource`. All subclasses are loaded in the `__init__.py` of the `data_source` submodule. If this does not work, I raise a `ValueError`.
3. Then I call the Pydantic `BaseModel.model_validate` class method on the entry, so that the correct `DataSource` (e.g. `GeoDataFrameSource` is instantiated. This is also where the validation happens. The result is returned

Reply:
that makes sense, good job!